man/man5/rlm_attr_rewrite.5
.TH rlm_attr_rewrite 5 "3 February 2004" "" "FreeRADIUS Module"
.SH NAME
rlm_attr_rewrite \- FreeRADIUS Module
.SH DESCRIPTION
The \fIrlm_attr_rewrite\fP module provides functionality to rewrite
data in radius packets on the fly.  This can be useful in
authorization and accounting functions.
.PP
The main configuration items to be aware of are:
.IP attribute
The name of the attribute that you want to be rewritten
.IP searchfor
String to be replaced.
.IP searchin
Where to look for the attribute.  Valid values are:
.BR packet,
.BR reply,
.BR proxy,
.BR proxy_reply,
.BR config
.IP replacewith
String to replace "searchfor".
.IP append
Set to yes or no.  If set to yes then the replace string will be
appended to the original string.
.IP ignore_case
Set to yes or no.  Determines whether search will be case insensitive
or case sensitive.
.IP new_attribute
Set to yes or no.  If set to yes, then new attribute will be created
with values from "searchin" and "replacewith".  Other options will be
ignored.
.IP max_matches
Set the max number of attributes to match against.  This is useful if
you need to rewrite the value in multiple Vendor Specific Attributes.
.PP
.SH SECTIONS
.BR authorization,
.BR authentication,
.BR postauthentication,
.BR preaccounting,
.BR accounting,
.BR preproxy,
.BR postproxy
.PP
.SH FILES
.I /etc/raddb/radiusd.conf
.PP
.SH "SEE ALSO"
.BR radiusd (8),
.BR radiusd.conf (5)
.SH AUTHOR
Chris Parker, cparker@segv.org



doc/ChangeLog
FreeRADIUS 3.2.11 Wed 03 Jun 2026 12:00:00 UTC urgency=low
	Configuration changes
	* PEAP / SoH functionality has been removed.  It was removed from
	  Windows over 10 years ago, and is no longer relevant.
	* Add example of using socat to test radclient + UDP to the server
	  over TLS.  See sites-available/tls, and look for "socat".
	* Add per-section "do_xlat" and "is_json" config items to rlm_rest.
	  see mods-available/rest for details.
	* Add notes on the security of CHAP-Password and MS-CHAP-*
	  If you send these attributes over RADIUS/UDP, then any observer
	  can crack those passwords in milliseconds on a laptop.
	* Removed all code and configuration related to trustrouter (abfab)
	  That project was marked as end of life a few years ago.

	Feature improvements
	* Implement "track_home_server" in home_server_pool configuration
	  See proxy.conf.  The previous (and not working) "affinity_id"
	  configuration has been removed.
	* Implement EAP-PSK.  See mods-available/eap for configuration.
	  The PSK must be in "PSK-Password", and must be 16 octets in length
	* Add $VALUE{filename} which reads a value from one line of a
	  file.
	* Add $FILE{filename} which reads a value from the entire contents
	  of a file.
	* Update DPSK module to read VLANs from the CSV file, and to export
	  &reply:Pairwise-Master-Key
	* Clean up and normalize log messages and debug output.  These
	  changes make the messages easier to read, and easier to understand.
	* Add significantly more stats for outgoing proxy sockets.

	Bug fixes
	* Skip TLS PSK checks when OpenSSL does not support PSK.
	  Fixes #5874.
	* Updated to help build with OpenSSL 4.0 alpha.
	* Improve performance of State handling when tracking many EAP packets.
	* Fix overflow on relayed DHCP packets.  Issue found by A. Ramos.
	* Fix more corner cases in RadSec when sockets have errors.
	* Tweaked module messages to be more consistent, so that messages
	  associated with a request are printed with the correct request ID.
	* Fix race conditions when adding or deleting multiple dynamic TLS clients.
	* Ethernet addresses which are cast to int64 are now in the low bits of
	  the number, instead of the high bits.
	* Remove redundant talloc_free in LDAP.  Fixes #5927.
	* Add printing of prepend operator.  Fixes #5928.
	* Failed TCP / TLS connections no longer mark the home server as "dead".
	  Existing connections (if any) continue to be used.  New connection are
	  blocked for "connect_fail_interval" or "certificate_fail_interval".

FreeRADIUS 3.2.10 Wed 03 Jun 2026 12:00:00 UTC urgency=high
	Configuration changes

	Feature improvements

	Bug fixes
	* Correct bug where detail file reader would not read files.
	  Patch from Bjørn Mork.  Fixes #5870

FreeRADIUS 3.2.9 Mon 01 Jun 2026 12:00:00 UTC urgency=low
	Configuration changes
	* Add "protocol_error = yes" configuration to clients.  If set,
	  the server can return Protocol-Error responses to the client.
	* radclient can now suppress Message-Authenticator in Access-Request,
	  when the input packet contains "Message-Authenticator !* ANY"
	  Don't use this in production!
	* Set "suppress_secrets = true" by default.
	* Add "connect_fail_interval" to home_server configuration.  If a
	  connection fails, the server will wait this time before trying to
	  connect again.
	* Add "certificate_fail_interval" to home_server configuration.
	  If a connection succeeds but the home_server certificate is
	  invalid, the server will wait this time before trying to connect
	  again.
	* Add `update` section to home_server configuration.
	  Status-Server packets can therefore be customized.
	* Add "cipher_suites" to tls{} configuration.  See
	  raddb/sites-available/tls.  This is mainly used to set the
	  cipher suites for TLS-PSK with TLS 1.3.

	Feature improvements
	* Initial implementation of Protocol-Failure as per IETF draft.
	  The functionality is disabled by default, but can be enabled
	  via new configuration flags.
	* Always allow Protocol-Error packet as valid response to any
	  packet.
	* Add Error-Cause attributes to CoA-NAK and Disconnect-NAK
	* added filter_username_nai to policy.d/filter, mainly for use in
	  eduroam.
	* Updates to VSCode default configuration.
	* Cleanups and add log messages for rlm_proxy_rate_limit.
	* allow 389ds legacy PBKDF2_SHA256 to use arbitrary iteration
	  count.  (#5654)
	* amend policy insert_acct_class/acct_unique to work in
	  environments with multiple Class attributes (#5337)
	* Tweak sqlippool messages to make them clearer.
	* Print log message if the server receives a correct authenticated
	  proxy response packet, but which has an unexpected code.
	  e.g. received Access-Accept in response to an Accounting-Request.
	* New installations now set "suppress_secrets=true" by default.
	  The server also prints messages in debug mode which explains why
	  the secrets are being suppressed.
	* Allow parallel build for Debian.  Fixes #5774.
	* Add RTBrick and other dictionaries.
	* Add documentation for ntlm_auth and spaces in passwords.
	  Addresses #5654.

	Bug fixes
	* Many minor bug fixes and cleanups.
	* Fixes to RadSec.
	* Many other fixes to socket and event handling, which enable
	  increased scalability.
	* Fix issues found with EAP-MSCHAPv2, EAP-PWD, and EAP-MD5.
	* Fix run_dir (#5637) and MemoryLimit (#5639)
	* Disable the PCRE JIT at run time if it can't allocate
	  executable memory.
	* Set selinux boolean to allow PCRE2 JIT
	* If you set the clock 25 years in the future, don't spam
	  systemd.  Fixes #5642
	* Don't load the OpenSSL legacy provider when built with
	  --enable-fips-workaround.  Fixes #5644.
	* Address potential leaks when opening many RADIUS/TLS proxy
	  sockets.
	* Encode multiple DHCP Option 82 as one option, instead of as
	  multiple options.
	* Update the rlm_cache_redis driver to reconnect on connection
	  failure.  Fixes #5651.
	* Tweaks to the processing state machine to handle more corner

src/modules/rlm_attr_rewrite/rlm_attr_rewrite.c
/*
 * rlm_attr_rewrite.c
 *
 * Version:  $Id$
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 *
 * Copyright 2002,2006  The FreeRADIUS server project
 * Copyright 2002  Kostas Kalevras <kkalev@noc.ntua.gr>
 */

#include <freeradius-devel/ident.h>
RCSID("$Id$")

#include <freeradius-devel/radiusd.h>
#include <freeradius-devel/modules.h>

#ifdef HAVE_REGEX_H
#	include <regex.h>
#endif

#define RLM_REGEX_INPACKET 0
#define RLM_REGEX_INCONFIG 1
#define RLM_REGEX_INREPLY  2
#define RLM_REGEX_INPROXY 3
#define RLM_REGEX_INPROXYREPLY 4

typedef struct rlm_attr_rewrite_t {
	char *attribute;	/* The attribute to search for */
	int  attr_num;		/* The attribute number */
	char *search;		/* The pattern to search for */
	int search_len;		/* The length of the search pattern */
	char *searchin_str;	/* The VALUE_PAIR list to search in. Can be either packet,reply,proxy,proxy_reply or config */
	char searchin;		/* The same as above just coded as a number for speed */
	char *replace;		/* The replacement */
	int replace_len;	/* The length of the replacement string */
	int  append;		/* Switch to control append mode (1,0) */
	int  nocase;		/* Ignore case */
	int  new_attr;		/* Boolean. Do we create a new attribute or not? */
	int  num_matches;	/* Maximum number of matches */
	char *name;		/* The module name */
} rlm_attr_rewrite_t;

static const CONF_PARSER module_config[] = {
  { "attribute", PW_TYPE_STRING_PTR, offsetof(rlm_attr_rewrite_t,attribute), NULL, NULL },
  { "searchfor", PW_TYPE_STRING_PTR, offsetof(rlm_attr_rewrite_t,search), NULL, NULL },
  { "searchin",  PW_TYPE_STRING_PTR, offsetof(rlm_attr_rewrite_t,searchin_str), NULL, "packet" },
  { "replacewith", PW_TYPE_STRING_PTR, offsetof(rlm_attr_rewrite_t,replace), NULL, NULL },
  { "append", PW_TYPE_BOOLEAN, offsetof(rlm_attr_rewrite_t,append),NULL, "no" },
  { "ignore_case", PW_TYPE_BOOLEAN, offsetof(rlm_attr_rewrite_t,nocase), NULL, "yes" },
  { "new_attribute", PW_TYPE_BOOLEAN, offsetof(rlm_attr_rewrite_t,new_attr), NULL, "no" },
  { "max_matches", PW_TYPE_INTEGER, offsetof(rlm_attr_rewrite_t,num_matches), NULL, "10" },
  { NULL, -1, 0, NULL, NULL }
};

static int attr_rewrite_instantiate(CONF_SECTION *conf, void **instance)
{
	rlm_attr_rewrite_t *data;
	DICT_ATTR *dattr;
	const char *instance_name = NULL;

	/*
	 *	Set up a storage area for instance data
	 */
	data = rad_malloc(sizeof(*data));
	if (!data) {
		return -1;
	}
	memset(data, 0, sizeof(*data));

	/*
	 *	If the configuration parameters can't be parsed, then
	 *	fail.
	 */
	if (cf_section_parse(conf, data, module_config) < 0) {
		free(data);
		return -1;
	}

	/*
	 *	Discover the attribute number of the key.
	 */
	if (data->attribute == NULL) {
		radlog(L_ERR, "rlm_attr_rewrite: 'attribute' must be set.");
		return -1;
	}
	if (data->search == NULL || data->replace == NULL) {
		radlog(L_ERR, "rlm_attr_rewrite: search/replace strings must be set.");
		return -1;
	}
	data->search_len = strlen(data->search);
	data->replace_len = strlen(data->replace);

	if (data->replace_len == 0 && data->new_attr){
		radlog(L_ERR, "rlm_attr_rewrite: replace string must not be zero length in order to create new attribute.");
		return -1;
	}

	if (data->num_matches < 1 || data->num_matches > MAX_STRING_LEN) {
		radlog(L_ERR, "rlm_attr_rewrite: Illegal range for match number.");
		return -1;
	}
	if (data->searchin_str == NULL) {
		radlog(L_ERR, "rlm_attr_rewrite: Illegal searchin directive given. Assuming packet.");
		data->searchin = RLM_REGEX_INPACKET;
	}
	else{
		if (strcmp(data->searchin_str, "packet") == 0)
			data->searchin = RLM_REGEX_INPACKET;
		else if (strcmp(data->searchin_str, "config") == 0)
			data->searchin = RLM_REGEX_INCONFIG;
		else if (strcmp(data->searchin_str, "reply") == 0)
			data->searchin = RLM_REGEX_INREPLY;
		else if (strcmp(data->searchin_str, "proxy") == 0)
			data->searchin = RLM_REGEX_INPROXY;
		else if (strcmp(data->searchin_str, "proxy_reply") == 0)
			data->searchin = RLM_REGEX_INPROXYREPLY;
		else {
			radlog(L_ERR, "rlm_attr_rewrite: Illegal searchin directive given. Assuming packet.");
			data->searchin = RLM_REGEX_INPACKET;
		}
		free((char *)data->searchin_str);
	}
	dattr = dict_attrbyname(data->attribute);
	if (dattr == NULL) {
		radlog(L_ERR, "rlm_attr_rewrite: No such attribute %s",
				data->attribute);
		return -1;
	}
	data->attr_num = dattr->attr;
	/* Add the module instance name */
	data->name = NULL;
	instance_name = cf_section_name2(conf);
	if (instance_name != NULL)
		data->name = strdup(instance_name);


	*instance = data;

	return 0;
}

static int do_attr_rewrite(void *instance, REQUEST *request)
{
	rlm_attr_rewrite_t *data = (rlm_attr_rewrite_t *) instance;
	int ret = RLM_MODULE_NOOP;
	VALUE_PAIR *attr_vp = NULL;
	VALUE_PAIR *tmp = NULL;
	regex_t preg;
	regmatch_t pmatch[9];
	int cflags = 0;
	int err = 0;
	char done_xlat = 0;
	unsigned int len = 0;
	char err_msg[MAX_STRING_LEN];
	unsigned int i = 0;
	unsigned int j = 0;
	unsigned int counter = 0;
	char new_str[MAX_STRING_LEN];
	char *ptr, *ptr2;
	char search_STR[MAX_STRING_LEN];
	char replace_STR[MAX_STRING_LEN];
	int replace_len = 0;

	if ((attr_vp = pairfind(request->config_items, PW_REWRITE_RULE)) != NULL){
		if (data->name == NULL || strcmp(data->name,attr_vp->vp_strvalue))
			return RLM_MODULE_NOOP;
	}

	if (data->new_attr){
		/* new_attribute = yes */
		if (!radius_xlat(replace_STR, sizeof(replace_STR), data->replace, request, NULL)) {
			DEBUG2("rlm_attr_rewrite: xlat on replace string failed.");
			return ret;
		}
		replace_len = strlen(replace_STR);
		attr_vp = pairmake(data->attribute,replace_STR,0);
		if (attr_vp == NULL){
			DEBUG2("rlm_attr_rewrite: Could not add new attribute %s with value '%s'",
				data->attribute,replace_STR);
			return ret;
		}
		switch(data->searchin){
			case RLM_REGEX_INPACKET:
				pairadd(&request->packet->vps,attr_vp);
				break;
			case RLM_REGEX_INCONFIG:
				pairadd(&request->config_items,attr_vp);
				break;
			case RLM_REGEX_INREPLY:
				pairadd(&request->reply->vps,attr_vp);
				break;
			case RLM_REGEX_INPROXY:
				if (!request->proxy) {
					pairbasicfree(attr_vp);
					return RLM_MODULE_NOOP;
				}
				pairadd(&request->proxy->vps, attr_vp);
				break;
			case RLM_REGEX_INPROXYREPLY:
				if (!request->proxy_reply) {
					pairbasicfree(attr_vp);
					return RLM_MODULE_NOOP;
				}
				pairadd(&request->proxy_reply->vps, attr_vp);
				break;
			default:
				radlog(L_ERR, "rlm_attr_rewrite: Illegal value for searchin. Changing to packet.");
				data->searchin = RLM_REGEX_INPACKET;
				pairadd(&request->packet->vps,attr_vp);
				break;
		}
		DEBUG2("rlm_attr_rewrite: Added attribute %s with value '%s'",data->attribute,replace_STR);
		ret = RLM_MODULE_OK;
	} else {
		/* new_attribute = no */
		switch (data->searchin) {
			case RLM_REGEX_INPACKET:
				if (data->attr_num == PW_USER_NAME)
					attr_vp = request->username;
				else if (data->attr_num == PW_USER_PASSWORD)
					attr_vp = request->password;
				else
					tmp = request->packet->vps;
				break;
			case RLM_REGEX_INCONFIG:
				tmp = request->config_items;
				break;
			case RLM_REGEX_INREPLY:
				tmp = request->reply->vps;
				break;
			case RLM_REGEX_INPROXYREPLY:
				if (!request->proxy_reply)
					return RLM_MODULE_NOOP;
				tmp = request->proxy_reply->vps;
				break;
			case RLM_REGEX_INPROXY:
				if (!request->proxy)
					return RLM_MODULE_NOOP;
				tmp = request->proxy->vps;
				break;
			default:
				radlog(L_ERR, "rlm_attr_rewrite: Illegal value for searchin. Changing to packet.");
				data->searchin = RLM_REGEX_INPACKET;
				attr_vp = pairfind(request->packet->vps, data->attr_num);
				break;
		}
do_again:
		if (tmp != NULL)
			attr_vp = pairfind(tmp, data->attr_num);
		if (attr_vp == NULL) {
			DEBUG2("rlm_attr_rewrite: Could not find value pair for attribute %s",data->attribute);
			return ret;
		}
		if (attr_vp->vp_strvalue == NULL || attr_vp->length == 0){
			DEBUG2("rlm_attr_rewrite: Attribute %s string value NULL or of zero length",data->attribute);
			return ret;
		}
		cflags |= REG_EXTENDED;
		if (data->nocase)
			cflags |= REG_ICASE;

		if (!radius_xlat(search_STR, sizeof(search_STR), data->search, request, NULL) && data->search_len != 0) {
			DEBUG2("rlm_attr_rewrite: xlat on search string failed.");
			return ret;
		}

		if ((err = regcomp(&preg,search_STR,cflags))) {
			regerror(err, &preg, err_msg, MAX_STRING_LEN);
			DEBUG2("rlm_attr_rewrite: regcomp() returned error: %s",err_msg);
			return ret;
		}
		
		if ((attr_vp->type == PW_TYPE_IPADDR) &&
		    (attr_vp->vp_strvalue[0] == '\0')) {
		  inet_ntop(AF_INET, &(attr_vp->lvalue),
			    attr_vp->vp_strvalue,
			    sizeof(attr_vp->vp_strvalue));
		}

		ptr = new_str;
		ptr2 = attr_vp->vp_strvalue;
		counter = 0;

		for ( i = 0 ;i < (unsigned)data->num_matches; i++) {
			err = regexec(&preg, ptr2, REQUEST_MAX_REGEX, pmatch, 0);
			if (err == REG_NOMATCH) {
				if (i == 0) {
					DEBUG2("rlm_attr_rewrite: No match found for attribute %s with value '%s'",
							data->attribute, attr_vp->vp_strvalue);
					regfree(&preg);
					goto to_do_again;
				} else
					break;
			}
			if (err != 0) {
				regfree(&preg);
				radlog(L_ERR, "rlm_attr_rewrite: match failure for attribute %s with value '%s'",
						data->attribute, attr_vp->vp_strvalue);
				return ret;
			}
			if (pmatch[0].rm_so == -1)
				break;
			len = pmatch[0].rm_so;
			if (data->append) {
				len = len + (pmatch[0].rm_eo - pmatch[0].rm_so);
			}
			counter += len;
			if (counter >= MAX_STRING_LEN) {
				regfree(&preg);
				DEBUG2("rlm_attr_rewrite: Replacement out of limits for attribute %s with value '%s'",
						data->attribute, attr_vp->vp_strvalue);
				return ret;
			}

			strlcpy(ptr, ptr2,len);
			ptr += len;
			ptr2 += pmatch[0].rm_eo;

			if (i == 0){
				/*
				 * We only run on the first match, sorry
				 */
				for(j = 0; j <= REQUEST_MAX_REGEX; j++){
					char *p;
					char buffer[sizeof(attr_vp->vp_strvalue)];

					/*
				   	 * Stolen from src/main/valuepair.c, paircompare()
				 	 */

					/*
					 * Delete old matches if the corresponding match does not
					 * exist in the current regex
					 */
					if (pmatch[j].rm_so == -1){
						p = request_data_get(request,request,REQUEST_DATA_REGEX | j);
						if (p){
							free(p);
							continue;
						}
						break;
					}
					memcpy(buffer,
					       attr_vp->vp_strvalue + pmatch[j].rm_so,
					       pmatch[j].rm_eo - pmatch[j].rm_so);
					buffer[pmatch[j].rm_eo - pmatch[j].rm_so] = '\0';
					p = strdup(buffer);
					request_data_add(request,request,REQUEST_DATA_REGEX | j,p,free);
				}
			}

			if (!done_xlat){
				if (data->replace_len != 0 &&
				radius_xlat(replace_STR, sizeof(replace_STR), data->replace, request, NULL) == 0) {
					DEBUG2("rlm_attr_rewrite: xlat on replace string failed.");
					return ret;
				}
				replace_len = (data->replace_len != 0) ? strlen(replace_STR) : 0;
				done_xlat = 1;
			}

			counter += replace_len;
			if (counter >= MAX_STRING_LEN) {
				regfree(&preg);
				DEBUG2("rlm_attr_rewrite: Replacement out of limits for attribute %s with value '%s'",
						data->attribute, attr_vp->vp_strvalue);
				return ret;
			}
			if (replace_len){
				strlcpy(ptr, replace_STR, replace_len);
				ptr += replace_len;
			}
		}
		regfree(&preg);
		len = strlen(ptr2) + 1;		/* We add the ending NULL */
		counter += len;
		if (counter >= MAX_STRING_LEN){
			DEBUG2("rlm_attr_rewrite: Replacement out of limits for attribute %s with value '%s'",
					data->attribute, attr_vp->vp_strvalue);
			return ret;
		}
		strlcpy(ptr, ptr2, len);

		DEBUG2("rlm_attr_rewrite: Changed value for attribute %s from '%s' to '%s'",
				data->attribute, attr_vp->vp_strvalue, new_str);
		if (pairparsevalue(attr_vp, new_str) == NULL) {
			DEBUG2("rlm_attr_rewrite: Could not write value '%s' into attribute %s: %s", new_str, data->attribute, librad_errstr);
			return ret;
		}

to_do_again:
		ret = RLM_MODULE_OK;

		if (tmp != NULL){
			tmp = attr_vp->next;
			if (tmp != NULL)
				goto do_again;
		}
	}

	return ret;
}

static int attr_rewrite_accounting(void *instance, REQUEST *request)
{
	return do_attr_rewrite(instance, request);
}

static int attr_rewrite_authorize(void *instance, REQUEST *request)
{
	return do_attr_rewrite(instance, request);
}

static int attr_rewrite_authenticate(void *instance, REQUEST *request)
{
	return do_attr_rewrite(instance, request);
}

static int attr_rewrite_preacct(void *instance, REQUEST *request)
{
	return do_attr_rewrite(instance, request);
}

static int attr_rewrite_checksimul(void *instance, REQUEST *request)
{
	return do_attr_rewrite(instance, request);
}

static int attr_rewrite_preproxy(void *instance, REQUEST *request)
{
	return do_attr_rewrite(instance, request);
}

static int attr_rewrite_postproxy(void *instance, REQUEST *request)
{
	return do_attr_rewrite(instance, request);
}

static int attr_rewrite_postauth(void *instance, REQUEST *request)
{
	return do_attr_rewrite(instance, request);
}

static int attr_rewrite_detach(void *instance)
{
	rlm_attr_rewrite_t *data = (rlm_attr_rewrite_t *) instance;

	if (data->name)
		free(data->name);

	free(instance);
	return 0;
}

/*
 *	The module name should be the only globally exported symbol.
 *	That is, everything else should be 'static'.
 *
 *	If the module needs to temporarily modify it's instantiation
 *	data, the type should be changed to RLM_TYPE_THREAD_UNSAFE.
 *	The server will then take care of ensuring that the module
 *	is single-threaded.
 */
module_t rlm_attr_rewrite = {
	RLM_MODULE_INIT,
	"attr_rewrite",
	RLM_TYPE_THREAD_UNSAFE,		/* type */
	attr_rewrite_instantiate,		/* instantiation */
	attr_rewrite_detach,			/* detach */
	{
		attr_rewrite_authenticate,	/* authentication */
		attr_rewrite_authorize, 	/* authorization */
		attr_rewrite_preacct,		/* preaccounting */
		attr_rewrite_accounting,	/* accounting */
		attr_rewrite_checksimul,	/* checksimul */
		attr_rewrite_preproxy,		/* pre-proxy */
		attr_rewrite_postproxy,		/* post-proxy */
		attr_rewrite_postauth		/* post-auth */
	},
};
